[XGrid] Fix calculation of the scroll position in `scrollToIndexes`

m4theushw opened the pull request:
Fixes #1818
Preview: https://codesandbox.io/s/material-uix-gridscroll-to-indexes-demo-simple-forked-0oi0o?file=/src/App.js

This is an attempt to fix the related bug. It took me some good time to understand how the virtualization works. :)

From my investigation, I found that the amount of scroll needed to make a row visible is less than what I would normally expect (rowIndex * rowHeight). Take as example the top position of the row at the index 5 below. In theory, its position is 250px but I had to scroll just 200px to see it. The same occurs with `scrollToIndexes`. It was using the viewport height, which might include space for partially visible rows. Now it's using the same height that triggers the page change in the virtualization.

<img width="500" src="https://user-images.githubusercontent.com/42154031/121625547-08dca280-ca4a-11eb-9994-ed4bc914670e.gif" alt="Mfl74Wvk3z" style="max-width:100%;">

DanailH commented:
Not sure if it ready for review but I can see 3 main problems:
 1. When navigating with the keyboard the grids start to scroll immediately, not when the last visible is focused and you want to move to the next one that is not visible,
 2. When you scroll the grid the normal way and then select a row there is a "jump" that positions the cell you selected to the top edge of the viewport.
 3. The scroll to bottom using the keyboard shortcut seems to not working.

dtassone commented on packages/grid/_modules_/grid/hooks/features/virtualization/useGridVirtualRows.ts:
```diff
@@ -222,32 +220,23 @@ export const useGridVirtualRows = (apiRef: GridApiRef): void => {
-          logger.debug(`Row is below, setting top to ${scrollCoordinates.top}`);
+        const scrollPosition = currentRowPage * gridState.containerSizes!.renderingZoneScrollHeight;
+
+        if (windowRef.current!.scrollTop !== scrollPosition) {
```
you just want to scroll when you can see the cell as @DanailH mentioned

m4theushw commented on packages/grid/_modules_/grid/hooks/features/virtualization/useGridVirtualRows.ts:
```diff
@@ -233,10 +233,14 @@ export const useGridVirtualRows = (apiRef: GridApiRef): void => {
-        isRowIndexBelow = scrollPosition + rowHeight > windowRef.current!.scrollTop + pageHeight;
+        isRowIndexBelow =
+          scrollPosition + rowHeight >
+          windowRef.current!.scrollTop + viewportHeight + partialRowHeight;
```
Fixes a bug where the grid wasn't scrolling if the row was partially visible.

<img width="500" src="https://user-images.githubusercontent.com/42154031/121949061-6f094400-cd2e-11eb-9815-5601d2b1f170.gif" alt="BPYpligb7o" style="max-width:100%;">

m4theushw commented:
> When you scroll the grid the normal way and then select a row there is a "jump" that positions the cell you selected to the top edge of the viewport.

@DanailH Yeah, this was in the original CodeSandbox provided by the user. I removed it now.

> The scroll to the bottom using the keyboard shortcut seems to not working.

Fixed. I think it's ready for review now.

dtassone commented on packages/grid/_modules_/grid/hooks/features/virtualization/useGridVirtualRows.ts:
```diff
@@ -220,23 +222,33 @@ export const useGridVirtualRows = (apiRef: GridApiRef): void => {
-          logger.debug(`Setting top to ${scrollCoordinates.top}`);
-          scrollCoordinates.top = scrollPosition;
-          needScroll = true;
+        const viewportHeight = gridState.containerSizes!.renderingZoneScrollHeight;
```
it's a misleading var name I think

m4theushw replied:
True, I changed to `height`. Probably I'll revisit this hook to fix the jump, then we can discuss that again.

oliviertassinari commented on packages/grid/_modules_/grid/hooks/features/virtualization/useGridVirtualRows.ts:
```diff
@@ -225,24 +225,28 @@ export const useGridVirtualRows = (apiRef: GridApiRef): void => {
+        const height = gridState.containerSizes!.renderingZoneScrollHeight;
+        const scrollPosition = currentRowPage * height;
+
+        // The amount of space taken by the first partially visible row
+        const partialRowHeight = windowRef.current!.scrollTop % rowHeight;
```
See the logic that we copied from WAI, https://github.com/mui-org/material-ui/blob/2c4dc994da164a53b253f03bcd2425a52655f175/packages/material-ui/src/useAutocomplete/useAutocomplete.js#L328-L348. It's working smoothly https://next.material-ui.com/components/autocomplete/ and is exactly the same case.

oliviertassinari replied:
Wouldn't we get this problem solved if we get the root cause of #1911?

m4theushw replied:
> Wouldn't we get this problem solved if we get the root cause of #1911?

Yes, but #1911 will require a lot of work. We could first push this one to solve the pain of the user that opened #1818.

Does this diagram help to understand the logic?

![scroll position](https://user-images.githubusercontent.com/42154031/122142629-3fd5fe00-ce26-11eb-8c10-beb1cd5dec66.png)

oliviertassinari replied:
> Does this diagram help to understand the logic?

I meant that the current logic on HAD already take care of this.

> We could first push this one to solve the pain of the user that opened #1818.

From what I understand, this solution would only work if we reverse the wrong behavior of #1911.

oliviertassinari replied:
> I meant that the current logic on HEAD already takes care of this.

Ah, no, right, my bad. We forgot to account for the height of the row 😆 

```diff
diff --git a/packages/grid/_modules_/grid/hooks/features/virtualization/useGridVirtualRows.ts b/packages/grid/_modules_/grid/hooks/features/virtualization/useGridVirtualRows.ts
index 10878a5a..a19b4626 100644
--- a/packages/grid/_modules_/grid/hooks/features/virtualization/useGridVirtualRows.ts
+++ b/packages/grid/_modules_/grid/hooks/features/virtualization/useGridVirtualRows.ts
@@ -235,7 +235,7 @@ export const useGridVirtualRows = (apiRef: GridApiRef): void => {
         const viewportHeight = gridState.viewportSizes.height;

         isRowIndexAbove = windowRef.current!.scrollTop > scrollPosition;
-        isRowIndexBelow = windowRef.current!.scrollTop + viewportHeight < scrollPosition;
+        isRowIndexBelow = windowRef.current!.scrollTop + viewportHeight < scrollPosition + rowHeight;

         if (isRowIndexAbove) {
           scrollCoordinates.top = scrollPosition; // We put it at the top of the page
```

oliviertassinari replied:
> But something is still off 😆

Found it, `viewportHeight` value is wrong too.

m4theushw replied:
But we don't use `viewportHeight` here. We use `renderingZoneScrollHeight` for virtualization, which I think it's not right value.

oliviertassinari reviewed (commented):
I don't think that we should fix scrollToIndexes, it seems to behave correctly. What I think we should double down on is: https://github.com/mui-org/material-ui-x/issues/1911#issuecomment-861914182

oliviertassinari commented:
@m4theushw Is there something left to do on this PR?

m4theushw commented:
@oliviertassinari I've simplified the logic. We can push this one or wait for the virtualization logic to be redone.

oliviertassinari commented on packages/grid/x-grid/src/tests/rows.XGrid.test.tsx:
```diff
@@ -449,6 +449,35 @@ describe('<XGrid /> - Rows', () => {
+        // Scroll to rowIndex=4, which is the 1st row in page 1
+        apiRef.current.scrollToIndexes({ rowIndex: 4, colIndex: 0 });
+        gridWindow.dispatchEvent(new Event('scroll'));
+        expect(gridWindow.scrollTop).to.equal(renderingZoneHeight);
```
This doesn't look right. AFAIK, it should be
```suggestion
        expect(gridWindow.scrollTop).to.equal(4 * rowHeight);
```

We display the 5th row, so we need to scroll 4-row heights. Anything else would push us toward changing the scroll logic when it's the rendering logic that is wrong.

m4theushw replied:
#1969 doesn't solve the bug when the given `rowIndex` is way before the first visible row. This test is lying.

oliviertassinari replied:
Which bug are we talking about? If we are talking about #1818, I think that it should be put on hold until #1911 is fixed.

m4theushw replied:
You mentioned that this PR passes with the logic from #1969. From what I understood that means that we should only merge the test case here. I'm saying that the bug of #1818 remains and the test case is not right. I'll close this one and investigate how to address #1911.

oliviertassinari replied:
- :+1: for an extra test case. 
- #1911 I would expect the solution not to be too complex. I think that we should first look into why the scrollTop is too small compared to what it should be, then once fixed, look into why the rendered rows are not continuous. It's likely a small refactoring.

oliviertassinari replied:
@m4theushw Coming back to a discussion we had. Do you agree that this assertion would be correct? https://github.com/mui-org/material-ui-x/pull/1871/files#r660182225. If you do, I think that we could merge a new test case. If you don't, we can chat more about it.

Either way, what's the intended outcome of keeping the PR open for a couple of weeks? We could restart the effort from a blank branch.

m4theushw replied:
I do agree that https://github.com/mui-org/material-ui-x/pull/1871/files#r660182225 is the correct assertion, but it doesn't work this way. The virtualization changes from page 0 to 1 before that (at the "jump"). If we merge this one it will be too tied to the virtualization implementation. Better to close it and to fix the root cause.

oliviertassinari replied:
I was going to say that the value of this test case is to increase our test coverage, but I feel that it duplicate with https://github.com/mui-org/material-ui-x/blob/4066078512087318ec22e2b80763564d67b0a22e/packages/grid/x-grid/src/tests/rows.XGrid.test.tsx#L480

So, after all, I don't see any value in this PR 😁, outside of the discussion it has created.

m4theushw replied:
I'm testing it after rebasing with master. It seems to pass with the proposed assertion.

m4theushw replied:
Yeah, it passes but visually it's wrong.

https://user-images.githubusercontent.com/42154031/124522376-6eae1700-ddc9-11eb-8014-16c907d76211.mp4

LorenGr commented:
I am seeing that this topic was closed. But the bug is still present. Is there a plan to fix this please?
I made a demo here : https://codesandbox.io/s/datagridprodemo-material-demo-forked-xhl3m?file=/demo.js:660-675
the scrolling looses synch with the selected row when you scroll down a lot

m4theushw commented:
@LorenGr The calculation is right, but the virtualization is rendering the wrong rows. You can subscribe to #2673 to get notified when the new virtualization engine gets released.
